File: backend/myPizza/src/main/java/backendapp/myPizza/controllers/AuthController.java
package backendapp.myPizza.controllers;

import backendapp.myPizza.Models.entities.User;
import backendapp.myPizza.Models.enums.TokenPairType;
import backendapp.myPizza.Models.reqDTO.LoginDTO;
import backendapp.myPizza.Models.reqDTO.UserPostDTO;
import backendapp.myPizza.Models.resDTO.ConfirmRes;
import backendapp.myPizza.Models.resDTO.TokenPair;
import backendapp.myPizza.exceptions.BadRequestException;
import backendapp.myPizza.exceptions.UnauthorizedException;
import backendapp.myPizza.services.AuthService;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;

import java.io.IOException;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.Map;

@RestController
@RequestMapping("/auth")
public class AuthController {
    @Autowired
    private AuthService authSvc;

    @PostMapping("/register")
    public User register(@RequestBody UserPostDTO userPostDTO) throws BadRequestException {
        return authSvc.register(userPostDTO);
    }





    @PostMapping("/login")
    public ConfirmRes login(@RequestBody LoginDTO loginDTO, HttpServletResponse res) throws UnauthorizedException, NoSuchAlgorithmException, InvalidKeySpecException, IOException {
        Map<TokenPairType, TokenPair> tokenMap = authSvc.login(loginDTO.email(), loginDTO.password());
        TokenPair httpTokens = tokenMap.get(TokenPairType.HTTP);
        TokenPair wsTokens = tokenMap.get(TokenPairType.WS);
        Cookie accessToken = new Cookie("__access_tkn", httpTokens.getAccessToken());
        accessToken.setHttpOnly(true);
        accessToken.setDomain("localhost");
        accessToken.setPath("/");
        Cookie refreshToken = new Cookie("__refresh_tkn", httpTokens.getRefreshToken());
        refreshToken.setHttpOnly(true);
        refreshToken.setDomain("localhost");
        refreshToken.setPath("/");
        Cookie wsAccessToken = new Cookie("__ws_access_tkn", wsTokens.getAccessToken());
        wsAccessToken.setHttpOnly(true);
        wsAccessToken.setDomain("localhost");
        wsAccessToken.setPath("/");
        Cookie wsRefreshToken = new Cookie("__ws_refresh_tkn", wsTokens.getRefreshToken());
        wsRefreshToken.setHttpOnly(true);
        wsRefreshToken.setDomain("localhost");
        wsRefreshToken.setPath("/");
        res.addCookie(accessToken);
        res.addCookie(refreshToken);
        res.addCookie(wsAccessToken);
        res.addCookie(wsRefreshToken);
        return new ConfirmRes("Authenticated successfully", HttpStatus.OK);
    }

    @GetMapping("/logout")
    public ConfirmRes logout(HttpServletResponse res) {
        Cookie accessToken = new Cookie("__access_tkn", null);
        Cookie refreshToken = new Cookie("__refresh_tkn", null);
        Cookie wsAccessToken = new Cookie("__ws_access_tkn", null);
        Cookie wsRefreshToken = new Cookie("__ws_refresh_tkn", null);
        accessToken.setMaxAge(0);
        refreshToken.setMaxAge(0);
        wsAccessToken.setMaxAge(0);
        wsRefreshToken.setMaxAge(0);
        accessToken.setPath("/");
        refreshToken.setPath("/");
        wsAccessToken.setPath("/");
        wsRefreshToken.setPath("/");
        res.addCookie(accessToken);
        res.addCookie(refreshToken);
        res.addCookie(wsAccessToken);
        res.addCookie(wsRefreshToken);
        return new ConfirmRes("Logout successfully", HttpStatus.OK);
    }

}

File: backend_nodejs_nest_fastify/my-pizza/src/nest_modules/auth-user/controllers/auth.controller.ts
import { Body, Controller, Get, HttpStatus, Logger, Post, Res } from '@nestjs/common';
import { UserPostDTO } from '../interfaces/user-post-dto.interface';
import { AuthService } from '../services/auth.service';
import { UserRes } from '../interfaces/user-res.interface';
import { ConfigService } from '@nestjs/config';
import { LoginDTO } from '../interfaces/login-dto.interface';
import { ConfirmRes } from '../interfaces/confirm-res.interface';
import { FastifyReply } from 'fastify';
import { TokenPairType } from '../enums/token-pair-type.enum';

@Controller('auth')
export class AuthController {

    constructor(private authSvc: AuthService, private configSvc: ConfigService) { }

    private logger = new Logger('AuthController')
    
    @Post('/register')
    public async register(@Body() userPostDTO: UserPostDTO): Promise<UserRes> {
        return await this.authSvc.register(userPostDTO)
    }

    @Post('/login')
    public async login(@Body() loginDTO: LoginDTO, @Res({ passthrough: true }) res: FastifyReply): Promise<ConfirmRes> {
        const tokenMap = await this.authSvc.login(loginDTO)
        // this.logger.log(tokenMap.get(TokenPairType.HTTP))
        res.statusCode = HttpStatus.OK
        if (loginDTO.restore) {
            res.setCookie('__access_tkn', tokenMap.get(TokenPairType.HTTP).accessToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true,
                maxAge: this.configSvc.get('EXP.refreshTokenExp')
            })
            res.setCookie('__refresh_tkn', tokenMap.get(TokenPairType.HTTP).refreshToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true,
                maxAge: this.configSvc.get('EXP.refreshTokenExp')
            })
            res.setCookie('__ws_access_tkn', tokenMap.get(TokenPairType.WS).accessToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true,
                maxAge: this.configSvc.get('EXP.refreshTokenExp')
            })
            res.setCookie('__ws_refresh_tkn', tokenMap.get(TokenPairType.WS).refreshToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true,
                maxAge: this.configSvc.get('EXP.refreshTokenExp')
            })
        } else {
            res.setCookie('__access_tkn', tokenMap.get(TokenPairType.HTTP).accessToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true
            })
            res.setCookie('__refresh_tkn', tokenMap.get(TokenPairType.HTTP).refreshToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true
            })
            res.setCookie('__ws_access_tkn', tokenMap.get(TokenPairType.WS).accessToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true
            })
            res.setCookie('__ws_refresh_tkn', tokenMap.get(TokenPairType.WS).refreshToken, {
                domain: this.configSvc.get('COOKIE.domain'),
                sameSite: this.configSvc.get('COOKIE.sameSite'),
                httpOnly: true,
                path: '/',
                secure: true
            })
        }
        return {
            statusCode: HttpStatus.OK,
            timestamp: new Date().getTime(),
            message: "Authenticated successfully"
        }
    }

    @Get('/logout')
    public logout(@Res({ passthrough: true }) res: FastifyReply): ConfirmRes {
        res.clearCookie('__access_tkn')
        res.clearCookie('__refresh_tkn')
        res.clearCookie('__ws_access_tkn')
        res.clearCookie('__ws_refresh_tkn')
        return {
            statusCode: HttpStatus.OK,
            timestamp: new Date().getTime(),
            message: "Logged out successfully"
        }
    }


}
